### PR TITLE
parser - Handle malformed wiki entries.

### PR DESCRIPTION
**Pull request by josepht**
* Split wiki into separate entries so that a YAML error in
  one document does not stop the parser from parsing the rest
  of the documents.

* Use a FakeServer for integration tests.

LP:#1599598

### CODE REVIEW

File: snapcraft/internal/parser.py [hunk 1]
```diff
@@ -216,16 +220,25 @@ def _process_index(output):
...
-            logger.warning('Invalid wiki entry: {}'.format(e))
-            continue
-
-        if is_valid_parts_list(parts_list, after_parts):
-            master_parts_list.update(parts_list)
+    # split the wiki into an array of entries to allow the parser to
+    # proceed when invalid yaml is found.
+    entries = output.split(b'\n---\n')
+    for entry in entries:
+        print("JOE: entry: {}".format(entry))
```
Comment (sergiusens) [inline snapcraft/internal/parser.py]: no :wink:

File: snapcraft/tests/fake_servers.py [hunk 2]
```diff
@@ -90,6 +90,38 @@ def do_GET(self):
...
+origin: https://github.com/sergiusens/curl.git
+project-part: curl
+description:
+  Description here
+maintainer: none
+"""
+        self.send_header('Content-Type', 'text/plain')
+        if 'NO_CONTENT_LENGTH' not in os.environ:
+            self.send_header('Content-Length', len(response.encode()))
+        self.send_header('ETag', '1111')
```
Comment (sergiusens) [inline snapcraft/tests/fake_servers.py]: do you use this?

File: snapcraft/internal/parser.py [hunk 3]
```diff
@@ -216,16 +220,25 @@ def _process_index(output):
...
-            parts_list, after_parts = _process_entry(data)
-        except InvalidEntryError as e:
-            logger.warning('Invalid wiki entry: {}'.format(e))
-            continue
-
-        if is_valid_parts_list(parts_list, after_parts):
-            master_parts_list.update(parts_list)
+    # split the wiki into an array of entries to allow the parser to
+    # proceed when invalid yaml is found.
+    entries = output.split(b'\n---\n')
```
Comment (sergiusens) [inline snapcraft/internal/parser.py]: not a super duper fan of this, but oh well.

**Comment by sergiusens** [comment]
Not related to this, but have you thought of using `requests` instead of urllib, there's a nifty thing where you can request the `.text` for the response and forget the binary encoding.

**Comment by kyrofa** [comment]
Yeah, `requests` is awesome.

File: snapcraft/internal/parser.py (hunk 1 above)
Comment (josepht) [inline snapcraft/internal/parser.py]: fixed

File: snapcraft/tests/fake_servers.py (hunk 2 above)
Comment (josepht) [inline snapcraft/tests/fake_servers.py]: No, removed.

File: snapcraft/internal/parser.py (hunk 3 above)
Comment (josepht) [inline snapcraft/internal/parser.py]: I'm not either.  I'm thinking I might see if I can build a generator that handles reading the lines of the wiki output and yields a part entry for each YAML document in the wiki.  Thoughts?

Comment (sergiusens) [inline snapcraft/internal/parser.py]: sounds good IMO

**Comment by josepht** [comment]
Coverage reports aren't being done for the integration tests, right?  I'm not currently using the FakePartsWiki server in anything but integration tests so there's some new code that coverage does not see as tested.

File: snapcraft/internal/parser.py (hunk 3 above)
Comment (josepht) [inline snapcraft/internal/parser.py]: I've added a bug https://bugs.launchpad.net/snapcraft/+bug/1599974 to track that change.

**Comment by josepht** [comment]
retest this please

**Comment by josepht** [comment]
retest this please

**Comment by josepht** [comment]
Can this be merged now?

**Pull request merged**